plugin/loop/setup.go

```diff
@@ -0,0 +1,90 @@
+package loop
+
+import (
+	"math/rand"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/dnsutil"
+
+	"github.com/mholt/caddy"
+)
+
+func init() {
+	caddy.RegisterPlugin("loop", caddy.Plugin{
+		ServerType: "dns",
+		Action:     setup,
+	})
+}
+
+func setup(c *caddy.Controller) error {
+	l, err := parse(c)
+	if err != nil {
+		return plugin.Error("loop", err)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		l.Next = next
+		return l
+	})
+
+	// Send query to ourselves and see if it end up with us again.
+	c.OnStartup(func() error {
+		// Another Go function, otherwise we block startup and can't send the packet.
+		go func() {
+			deadline := time.Now().Add(30 * time.Second)
+			conf := dnsserver.GetConfig(c)
+
+			for time.Now().Before(deadline) {
+				ok := 0
+				for _, lh := range conf.ListenHosts {
+					addr := net.JoinHostPort(lh, conf.Port)
+					if _, err := l.exchange(addr); err != nil {
+						continue
+					}
+					ok++
+				}
+
+				if ok == len(conf.ListenHosts) {
+					go func() {
```

Comment:
looks like a way to disable sooner if at least one msg sent per Listening host.
Because of the double loop (30 sec, and for each Host) - and because of 'continue" at line 46, we are not sure that all Host are tested. (some may be tested several time, some not at all).